kratos/processes/fast_transfer_between_model_parts_process.cpp

```diff
@@ -126,75 +119,79 @@ void FastTransferBetweenModelPartsProcess::TransferWithFlags()
             #pragma omp for
             for(int i = 0; i < num_nodes; ++i) {
-                auto it_node = mrOriginModelPart.NodesBegin() + i;
+                auto it_node = it_node_begin + i;
                 if (it_node->Is(mFlag)) {
-                    (nodes_buffer_vector[thread_id]).insert((nodes_buffer_vector[thread_id]).begin(), *(it_node.base()));
+                    nodes_buffer_vector.insert(nodes_buffer_vector.begin(), *(it_node.base()));
                 }
             }
         }
 
         if (num_elements != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::ELEMENTS || mEntity == EntityTransfered::NODESANDELEMENTS)) {
+            const auto it_elem_begin = mrOriginModelPart.ElementsBegin();
             #pragma omp for
             for(int i = 0; i < num_elements; ++i) {
-                auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+                auto it_elem = it_elem_begin + i;
                 if (it_elem->Is(mFlag)) {
-                    (elements_buffer_vector[thread_id]).insert((elements_buffer_vector[thread_id]).begin(), *(it_elem.base()));
+                    elements_buffer_vector.insert(elements_buffer_vector.begin(), *(it_elem.base()));
                 }
             }
         }
 
         if (num_conditions != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONDITIONS || mEntity == EntityTransfered::NODESANDCONDITIONS)) {
+            const auto it_cond_begin = mrOriginModelPart.ConditionsBegin();
             #pragma omp for
             for(int i = 0; i < num_conditions; ++i) {
-                auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+                auto it_cond = it_cond_begin + i;
                 if (it_cond->Is(mFlag)) {
-                    (conditions_buffer_vector[thread_id]).insert((conditions_buffer_vector[thread_id]).begin(), *(it_cond.base()));
+                    conditions_buffer_vector.insert(conditions_buffer_vector.begin(), *(it_cond.base()));
                 }
             }
         }
 
         if (num_constraints != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONSTRAINTS || mEntity == EntityTransfered::NODESANDCONSTRAINTS)) {
+            const auto it_const_begin = mrOriginModelPart.MasterSlaveConstraintsBegin();
             #pragma omp for
             for(int i = 0; i < num_constraints; ++i) {
-                auto it_const = mrOriginModelPart.MasterSlaveConstraintsBegin() + i;
+                auto it_const = it_const_begin + i;
                 if (it_const->Is(mFlag)) {
-                    (constraints_buffer_vector[thread_id]).insert((constraints_buffer_vector[thread_id]).begin(), *(it_const.base()));
+                    constraints_buffer_vector.insert(constraints_buffer_vector.begin(), *(it_const.base()));
                 }
             }
         }
 
         // We transfer
-        #pragma omp single
+        #pragma omp critical
```

Comment:
why multiple criticals? i think it would be better to have everything into one single critical statement

Reply:
@roigcarlo recommended that the last time